File: apps/jamnative/Jambot/Services/RenderCache.swift
import Foundation
import CryptoKit
import os

// On-device render cache — the native twin of src/app/jam/renderCache.ts.
//
// The last whole-track render of each track is kept under
// Caches/renders/<trackId>.pcm (the planar Int16 samples exactly as the
// engine handed them over: channel 0, then channel 1) with a
// <trackId>.json sidecar (key + the RenderResult metadata). The key is a
// SHA-256 of "<engine stamp>|<serialized session JSON>", so any change to
// the session — or an engine bundle update — misses the cache and a stale
// render can never be played for a different state. Reopening an unchanged
// track plays instantly instead of re-rendering (a long song takes tens of
// seconds on a phone). The 6 most recently saved tracks are kept.
//
// Nothing here throws: a cache problem is logged and treated as a miss.
// The file work runs on a utility queue so a 42 MB write (128 bars, stereo)
// never touches the main thread; reads go straight from Data into the
// [Int16] with one bulk copy.
final class RenderCache {
    static let log = Logger(subsystem: "com.bartdecrem.Jambot", category: "cache")
    static let shared = RenderCache()

    /// Tracks kept on disk (most recently saved first).
    static let keep = 6

    /// Where the files live; `nil` when the Caches directory is unavailable.
    let directory: URL?

    private let queue = DispatchQueue(label: "com.bartdecrem.Jambot.renderCache", qos: .utility)

    /// Sidecar metadata — everything in `RenderResult` but the samples,
    /// plus the key it was rendered for.
    struct Sidecar: Codable, Equatable {
        var key: String
        var bpm: Int
        var bars: Int
        var hasArrangement: Bool
        var message: String
        var sampleRate: Double
        var channels: Int
        var length: Int
        var savedAt: Double // seconds since 1970
    }

    init(directory: URL? = nil) {
        if let directory {
            self.directory = directory
        } else if let caches = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first {
            self.directory = caches.appendingPathComponent("renders", isDirectory: true)
        } else {
            self.directory = nil
        }
    }

    // MARK: - Key

    /// Cache key for a session state: SHA-256 hex of
    /// `"<stamp>|<sessionJSON>"` — same text as the web's `renderCacheKey`.
    /// `stamp` is the engine bundle stamp (`EngineHost.engineVersion`), so a
    /// bundle update re-renders even when the session did not change.
    static func key(sessionJSON: Data, stamp: String) -> String {
        var hasher = SHA256()
        hasher.update(data: Data("\(stamp)|".utf8))
        hasher.update(data: sessionJSON)
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }

    /// Convenience for callers holding the session as `JSONValue` (what
    /// `EngineAPI.serialize()` returns). Encoded with sorted keys so the
    /// same state always hashes the same.
    static func key(session: JSONValue, stamp: String) -> String? {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        guard let data = try? encoder.encode(session) else { return nil }
        return key(sessionJSON: data, stamp: stamp)
    }

    // MARK: - Read

    /// The cached whole-track render for `trackId` when its key matches,
    /// else `nil`. Never throws.
    func load(trackId: String, key: String) async -> RenderResult? {
        await withCheckedContinuation { cont in
            queue.async { cont.resume(returning: self.loadSync(trackId: trackId, key: key)) }
        }
    }

    private func loadSync(trackId: String, key: String) -> RenderResult? {
        guard let dir = directory else { return nil }
        let sidecarURL = dir.appendingPathComponent("\(Self.safe(trackId)).json")
        let pcmURL = dir.appendingPathComponent("\(Self.safe(trackId)).pcm")
        guard let sidecarData = try? Data(contentsOf: sidecarURL),
              let meta = try? JSONDecoder().decode(Sidecar.self, from: sidecarData) else { return nil }
        guard meta.key == key else {
            Self.log.notice("cache miss \(trackId, privacy: .public): key changed")
            return nil
        }
        let samples = meta.length * meta.channels
        guard samples > 0 else { return nil }
        let started = Date()
        // Map the file rather than read it — the copy below is the only pass.
        guard let data = try? Data(contentsOf: pcmURL, options: .mappedIfSafe),
              data.count == samples * MemoryLayout<Int16>.size else {
            Self.log.error("cache \(trackId, privacy: .public): pcm size mismatch, dropping")
            dropSync(trackId: trackId)
            return nil
        }
        // One bulk copy into the array — no per-sample work.
        let pcm = [Int16](unsafeUninitializedCapacity: samples) { dst, count in
            data.withUnsafeBytes { src in
                let bytes = dst.count * MemoryLayout<Int16>.size
                UnsafeMutableRawBufferPointer(dst).copyMemory(from: UnsafeRawBufferPointer(rebasing: src.prefix(bytes)))
            }
            count = samples
        }
        let ms = Int(Date().timeIntervalSince(started) * 1000)
        Self.log.notice("cache hit \(trackId, privacy: .public): \(meta.bars) bars, \(samples * 2 / 1_048_576) MB in \(ms) ms")
        // Touch the sidecar so pruning keeps what was actually used recently.
        try? FileManager.default.setAttributes([.modificationDate: Date()], ofItemAtPath: sidecarURL.path)
        return RenderResult(bars: meta.bars, bpm: meta.bpm, hasArrangement: meta.hasArrangement, message: meta.message,
                            sampleRate: meta.sampleRate, channels: meta.channels, length: meta.length, pcm: pcm)
    }

    // MARK: - Write

    /// Store a whole-track render and prune to the `keep` most recent
    /// tracks. Never throws. Only cache song-scope renders — a section
    /// audition is not "the track".
    func save(trackId: String, key: String, _ result: RenderResult) async {
        await withCheckedContinuation { cont in
            queue.async { self.saveSync(trackId: trackId, key: key, result); cont.resume() }
        }
    }

    private func saveSync(trackId: String, key: String, _ r: RenderResult) {
        guard let dir = directory else { return }
        guard r.length > 0, r.channels >= 1, r.pcm.count >= r.length * r.channels else {
            Self.log.error("cache save \(trackId, privacy: .public): malformed render, skipped")
            return
        }
        let fm = FileManager.default
        do {
            try fm.createDirectory(at: dir, withIntermediateDirectories: true)
            let base = Self.safe(trackId)
            let pcmURL = dir.appendingPathComponent("\(base).pcm")
            let sidecarURL = dir.appendingPathComponent("\(base).json")
            let samples = r.length * r.channels
            let started = Date()
            // Write the planar Int16 as-is — no conversion pass.
            let data = r.pcm.withUnsafeBufferPointer { buf in
                Data(buffer: UnsafeBufferPointer(rebasing: buf.prefix(samples)))
            }
            try data.write(to: pcmURL, options: .atomic)
            let meta = Sidecar(key: key, bpm: r.bpm, bars: r.bars, hasArrangement: r.hasArrangement, message: r.message,
                               sampleRate: r.sampleRate, channels: r.channels, length: r.length,
                               savedAt: Date().timeIntervalSince1970)
            try JSONEncoder().encode(meta).write(to: sidecarURL, options: .atomic)
            let ms = Int(Date().timeIntervalSince(started) * 1000)
            Self.log.notice("cache saved \(trackId, privacy: .public): \(r.bars) bars, \(data.count / 1_048_576) MB in \(ms) ms")
            pruneSync(keeping: base)
        } catch {
            Self.log.error("cache save \(trackId, privacy: .public) failed: \(error.localizedDescription, privacy: .public)")
        }
    }

    /// Keep the `keep` most recently saved tracks (by sidecar `savedAt`);
    /// `current` is never pruned.
    private func pruneSync(keeping current: String) {
        guard let dir = directory else { return }
        let fm = FileManager.default
        guard let files = try? fm.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil) else { return }
        var entries: [(base: String, savedAt: Double)] = []
        for url in files where url.pathExtension == "json" {
            let base = url.deletingPathExtension().lastPathComponent
            guard let data = try? Data(contentsOf: url), let meta = try? JSONDecoder().decode(Sidecar.self, from: data) else {
                // Unreadable sidecar: its pcm is unusable too.
                try? fm.removeItem(at: url)
                try? fm.removeItem(at: dir.appendingPathComponent("\(base).pcm"))
                continue
            }
            entries.append((base, base == current ? .greatestFiniteMagnitude : meta.savedAt))
        }
        // Orphan pcm files (crash between the two writes) go too.
        for url in files where url.pathExtension == "pcm" {
            let base = url.deletingPathExtension().lastPathComponent
            if !entries.contains(where: { $0.base == base }) { try? fm.removeItem(at: url) }
        }
        let stale = entries.sorted { $0.savedAt > $1.savedAt }.dropFirst(Self.keep)
        for e in stale {
            try? fm.removeItem(at: dir.appendingPathComponent("\(e.base).json"))
            try? fm.removeItem(at: dir.appendingPathComponent("\(e.base).pcm"))
            Self.log.notice("cache pruned \(e.base, privacy: .public)")
        }
    }

    // MARK: - Drop

    /// Remove a track's cached render (e.g. when the track is deleted).
    /// Never throws.
    func drop(trackId: String) {
        queue.async { self.dropSync(trackId: trackId) }
    }

    private func dropSync(trackId: String) {
        guard let dir = directory else { return }
        let base = Self.safe(trackId)
        try? FileManager.default.removeItem(at: dir.appendingPathComponent("\(base).json"))
        try? FileManager.default.removeItem(at: dir.appendingPathComponent("\(base).pcm"))
    }

    /// Track ids on disk right now (for About/debug and the smoke test).
    func cachedTrackIds() async -> [String] {
        await withCheckedContinuation { cont in
            queue.async {
                guard let dir = self.directory,
                      let files = try? FileManager.default.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil) else {
                    cont.resume(returning: []); return
                }
                cont.resume(returning: files.filter { $0.pathExtension == "json" }.map { $0.deletingPathExtension().lastPathComponent }.sorted())
            }
        }
    }

    /// Track ids are UUIDs today; keep the file name safe if that changes.
    private static func safe(_ id: String) -> String {
        String(id.unicodeScalars.map { CharacterSet.alphanumerics.contains($0) || $0 == "-" || $0 == "_" ? Character($0) : "_" })
    }
}

File: src/app/jam/renderCache.ts
// Render cache: the last whole-track render of each track, kept in IndexedDB
// as 16-bit PCM and keyed by a hash of the serialized session plus the engine
// bundle stamp. Reopening a track whose session has not changed plays from the
// cache instantly instead of re-rendering (a long song takes tens of seconds on
// a phone). Any change to the session changes the key, so a stale render can
// never be played for a different state; the engine stamp is in the key so a
// bundle update re-renders too.

import type { RenderResult } from './jambot'

const DB_NAME = 'jam-renders'
const STORE = 'renders'
const KEEP = 6            // most recent tracks kept per device

type Row = {
  trackId: string
  key: string
  bpm: number
  bars: number
  hasArrangement: boolean
  synths: string[]
  message: string
  sampleRate: number
  channels: number
  length: number
  pcm: ArrayBuffer       // planar Int16: channel 0 samples, then channel 1 …
  savedAt: number
}

function openDb(): Promise<IDBDatabase> {
  return new Promise((resolve, reject) => {
    if (typeof indexedDB === 'undefined') { reject(new Error('IndexedDB unavailable')); return }
    const req = indexedDB.open(DB_NAME, 1)
    req.onupgradeneeded = () => {
      const db = req.result
      if (!db.objectStoreNames.contains(STORE)) db.createObjectStore(STORE, { keyPath: 'trackId' })
    }
    req.onsuccess = () => resolve(req.result)
    req.onerror = () => reject(req.error || new Error('IndexedDB open failed'))
  })
}

function tx<T>(db: IDBDatabase, mode: IDBTransactionMode, run: (store: IDBObjectStore) => IDBRequest<T>): Promise<T> {
  return new Promise((resolve, reject) => {
    const t = db.transaction(STORE, mode)
    const req = run(t.objectStore(STORE))
    req.onsuccess = () => resolve(req.result)
    req.onerror = () => reject(req.error || new Error('IndexedDB request failed'))
  })
}

/** Cache key for a session: SHA-256 of its serialized JSON plus the engine stamp. */
export async function renderCacheKey(sessionJson: string, stamp: string): Promise<string> {
  const text = `${stamp}|${sessionJson}`
  try {
    const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(text))
    return Array.from(new Uint8Array(digest), (b) => b.toString(16).padStart(2, '0')).join('')
  } catch {
    // Insecure context (plain http on a LAN): FNV-1a is plenty for a change detector.
    let h = 0x811c9dc5
    for (let i = 0; i < text.length; i++) { h ^= text.charCodeAt(i); h = Math.imul(h, 0x01000193) >>> 0 }
    return `fnv-${h.toString(16)}-${text.length}`
  }
}

function makeBuffer(channels: number, length: number, sampleRate: number): AudioBuffer {
  try {
    return new AudioBuffer({ numberOfChannels: channels, length, sampleRate })
  } catch {
    const Ctor = (window as unknown as { OfflineAudioContext?: typeof OfflineAudioContext; webkitOfflineAudioContext?: typeof OfflineAudioContext }).OfflineAudioContext
      || (window as unknown as { webkitOfflineAudioContext?: typeof OfflineAudioContext }).webkitOfflineAudioContext
    if (!Ctor) throw new Error('no AudioBuffer constructor')
    return new Ctor(channels, length, sampleRate).createBuffer(channels, length, sampleRate)
  }
}

/** The cached render for a track if its key matches, else null. Never throws. */
export async function loadCachedRender(trackId: string, key: string): Promise<RenderResult | null> {
  try {
    const db = await openDb()
    const row = await tx<Row | undefined>(db, 'readonly', (s) => s.get(trackId) as IDBRequest<Row | undefined>)
    db.close()
    if (!row || row.key !== key || !row.pcm || !row.length) return null
    const buffer = makeBuffer(row.channels, row.length, row.sampleRate)
    const all = new Int16Array(row.pcm)
    for (let ch = 0; ch < row.channels; ch++) {
      const f32 = new Float32Array(row.length)
      const base = ch * row.length
      for (let i = 0; i < row.length; i++) f32[i] = all[base + i] / 32768
      buffer.copyToChannel(f32, ch)
    }
    return { buffer, message: row.message, bars: row.bars, bpm: row.bpm, synths: row.synths, hasArrangement: row.hasArrangement }
  } catch (e) {
    console.warn('[jam] render cache read failed', e)
    return null
  }
}

/** Store a whole-track render (planar Int16) and prune to the KEEP most recent tracks. Never throws. */
export async function saveRender(trackId: string, key: string, r: RenderResult): Promise<void> {
  try {
    const { buffer } = r
    const channels = buffer.numberOfChannels
    const length = buffer.length
    const all = new Int16Array(channels * length)
    for (let ch = 0; ch < channels; ch++) {
      const data = buffer.getChannelData(ch)
      const base = ch * length
      for (let i = 0; i < length; i++) {
        const v = Math.max(-1, Math.min(1, data[i]))
        all[base + i] = v < 0 ? v * 32768 : v * 32767
      }
    }
    const row: Row = {
      trackId, key, bpm: r.bpm, bars: r.bars, hasArrangement: r.hasArrangement, synths: r.synths, message: r.message,
      sampleRate: buffer.sampleRate, channels, length, pcm: all.buffer, savedAt: Date.now(),
    }
    const db = await openDb()
    await tx(db, 'readwrite', (s) => s.put(row))
    // Prune: keep the KEEP most recently saved tracks.
    const rows = await tx<Row[]>(db, 'readonly', (s) => s.getAll() as IDBRequest<Row[]>)
    const stale = rows.sort((a, b) => b.savedAt - a.savedAt).slice(KEEP)
    for (const old of stale) await tx(db, 'readwrite', (s) => s.delete(old.trackId))
    db.close()
  } catch (e) {
    console.warn('[jam] render cache write failed', e)
  }
}

/** Drop a track's cached render (e.g. when the track is deleted). Never throws. */
export async function dropCachedRender(trackId: string): Promise<void> {
  try {
    const db = await openDb()
    await tx(db, 'readwrite', (s) => s.delete(trackId))
    db.close()
  } catch { /* nothing to drop */ }
}
